### PR TITLE
Fix bug when adding additional sizes to empty size mapping.

### PR DESCRIPTION
**Pull request by domlander**
## What does this change?

- Fixes a bug where additional sizes could be added to a breakpoint which has no existing size definition. When adding additional sizes, instead of appending the additional sizes to the existing ad sizes from the next smallest breakpoint with ad sizes, it would overwrite these sizes.

## Why?

- Inline slots on liveblog pages at phablet & tablet viewport sizes were only using the additional sizes and not the sizes defined for mobile viewport with the additional sizes.

## Screenshots

| Before | After |
| - | - |
| ![Before] | ![After] |

[Before]: https://github.com/guardian/commercial/assets/9574885/283a6bdb-bcd4-43a1-9cb1-45304743a8f5
[After]: https://github.com/guardian/commercial/assets/9574885/ca2bfb84-0f80-442e-8bee-55fec6c1818c

### CODE REVIEW

File: src/core/ad-sizes.ts [hunk 1]
```diff
@@ -371,6 +382,38 @@ const slotSizeMappings: SlotSizeMappings = {
...
+ */
+const findAppliedSizesForBreakpoint = (
+	sizeMappings: SizeMapping,
+	breakpoint: Breakpoint,
+): AdSize[] => {
+	if (!isBreakpoint(breakpoint)) {
+		return [];
+	}
+
+	let breakpointIndex: 0 | 1 | 2 | 3 = breakpoints.findIndex(
```
Comment (domlander) [inline src/core/ad-sizes.ts:400]: Really not happy with this type. If we add another breakpoint to the `breakpoints` array we will need to update this type here. I couldn't find another way to get around the `noUncheckedIndexedAccess` typescript rule, so please shout if you have an idea!

Run `yarn tsc --project ./tsconfig.core.json` locally to test.

**Comment by Jakeii** [comment]
Rather than adding the complexity of `findAppliedSizesForBreakpoint` could we instead just have a typeguard to ensure that additionalSizes are only added for breakpoints that exist in the adsizes?

**Comment by domlander** [comment]
> Rather than adding the complexity of `findAppliedSizesForBreakpoint` could we instead just have a typeguard to ensure that additionalSizes are only added for breakpoints that exist in the adsizes? and restore that breakpoint that's missing and causing the bug.

We could use a typeguard. I don't see a way we could introduce a bug without a typeguard catching it. 

The reason I thought the function would be a good idea is so that if we want to add more additional sizes in the future, it will just work without having to consider the original definitions. If we had a type guard instead, we would have to duplicate a breakpoints size mappings whenever we wanted to add additional sizes. We'd also have to remember how the slot size mappings work (I've added a bunch of comments to help with this). That's why I thought a function would be better here.

File: src/core/ad-sizes.ts [hunk 2]
```diff
@@ -371,6 +385,38 @@ const slotSizeMappings: SlotSizeMappings = {
...
 const getAdSize = (size: SizeKeys): AdSize => adSizes[size];
 
+/**
+ * Finds the ad sizes that will be used for a breakpoint given a size mapping
+ *
+ * If ad sizes are defined in the size mapping for the specified breakpoint, we use that.
+ * If no sizes are defined for the breakpoint, use the next smallest breakpoint with defined ad sizes.
+ * If no smaller breakpoints have defined ad sizes, return an empty array
+ */
+const findAppliedSizesForBreakpoint = (
```
Comment (Jakeii) [inline src/core/ad-sizes.ts:395]: Could you maybe throw a couple of examples in the comment?

**Comment by Jakeii** [comment]
> > Rather than adding the complexity of `findAppliedSizesForBreakpoint` could we instead just have a typeguard to ensure that additionalSizes are only added for breakpoints that exist in the adsizes? and restore that breakpoint that's missing and causing the bug.
> 
> We could use a typeguard. I don't see a way we could introduce a bug without a typeguard catching it.
> 
> The reason I thought the function would be a good idea is so that if we want to add more additional sizes in the future, it will just work without having to consider the original definitions. If we had a type guard instead, we would have to duplicate a breakpoints size mappings whenever we wanted to add additional sizes. We'd also have to remember how the slot size mappings work (I've added a bunch of comments to help with this). That's why I thought a function would be better here.

I guess I was envisioning where we'd have a much stricter type of `slotSizeMapping` using `as const satisfies` or something. Allowing at compile time to know that you're trying to concat to a breakpoint (or slot) that does or doesn't exists.

But I've just had a quick experiment I think it'll probably require a pretty big refactor of the sizeMappings at best and might be impossible at worst!😅

File: src/core/ad-sizes.ts (hunk 2 above)
Comment (domlander) [reply src/core/ad-sizes.ts:409]: Will do. There's also more examples in the tests for this function

Comment (Jakeii) [reply src/core/ad-sizes.ts:409]: sweet

**Review by Jakeii — APPROVED** [review]

**Comment by Jakeii** [comment]
Great documentation improvements!

**Pull request merged**